internal/venice-common/src/main/java/com/linkedin/venice/pubsub/PubSubClientsFactory.java

```diff
@@ -30,4 +57,60 @@ public PubSubConsumerAdapterFactory getConsumerAdapterFactory() {
   public PubSubAdminAdapterFactory getAdminAdapterFactory() {
     return adminAdapterFactory;
   }
+
+  public static PubSubProducerAdapterFactory<PubSubProducerAdapter> createProducerFactory(Properties properties) {
+    return createProducerFactory(new VeniceProperties(properties));
+  }
+
+  public static PubSubProducerAdapterFactory<PubSubProducerAdapter> createProducerFactory(
+      VeniceProperties veniceProperties) {
+    return createFactory(
+        veniceProperties,
+        PUB_SUB_PRODUCER_ADAPTER_FACTORY_CLASS,
+        ApacheKafkaProducerAdapterFactory.class.getName(),
+        FactoryType.PRODUCER);
+  }
+
+  public static PubSubConsumerAdapterFactory<PubSubConsumerAdapter> createConsumerFactory(
+      VeniceProperties veniceProperties) {
+    return createFactory(
+        veniceProperties,
+        PUB_SUB_CONSUMER_ADAPTER_FACTORY_CLASS,
+        ApacheKafkaConsumerAdapterFactory.class.getName(),
+        FactoryType.CONSUMER);
+  }
+
+  public static PubSubAdminAdapterFactory<PubSubAdminAdapter> createAdminFactory(VeniceProperties veniceProperties) {
+    return createFactory(
+        veniceProperties,
+        PUB_SUB_ADMIN_ADAPTER_FACTORY_CLASS,
+        ApacheKafkaAdminAdapterFactory.class.getName(),
+        FactoryType.ADMIN);
+  }
+
+  private static <T> T createFactory(
+      VeniceProperties properties,
+      String configKey,
+      String defaultClassName,
+      FactoryType factoryType) {
+    String className;
+    if (properties.containsKey(configKey)) {
+      className = properties.getString(configKey);
+      LOGGER.info("Creating pub-sub {} adapter factory instance for class: {}", factoryType, className);
+    } else {
+      className = defaultClassName;
+      LOGGER.info("Creating pub-sub {} adapter factory instance with default class: {}", factoryType, className);
+    }
+
+    return createInstance(className);
+  }
+
+  public static <T> T createInstance(String className) {
+    try {
+      return (T) Class.forName(className).getDeclaredConstructor().newInstance();
+    } catch (Exception e) {
+      LOGGER.error("Failed to create instance of class: {}", className, e);
```

Comment:
Looks like duplicate? you are anyway throwing the same error below.

Reply:
Yeah that's good point. It's redundant if upper layers also log this exception